psycopg/psycopg/types/enum.py

```diff
@@ -0,0 +1,95 @@
+"""
+Adapters for the enum type.
+"""
+from enum import Enum
+from typing import Optional, TypeVar, Generic, Type
+
+from .string import StrBinaryDumper, StrDumper
+from .. import postgres
+from .._typeinfo import EnumInfo as EnumInfo  # exported here
+from ..abc import AdaptContext
+from ..adapt import Buffer, Loader
+from ..pq import Format
+
+E = TypeVar("E", bound=Enum)
+
+
+class EnumLoader(Loader, Generic[E]):
+    format = Format.TEXT
+    python_type: Type[E]
+
+    def __init__(self, oid: int, context: Optional[AdaptContext] = None):
+        super().__init__(oid, context)
+
+    def load(self, data: Buffer) -> E:
+        if isinstance(data, memoryview):
+            data = bytes(data)
+        return self.python_type(data.decode())
+
+
+class EnumBinaryLoader(EnumLoader[E]):
+    format = Format.BINARY
+
+
+class EnumDumper(StrDumper):
+    pass
+
+
+class EnumBinaryDumper(StrBinaryDumper):
+    pass
+
+
+def register_enum(
+    info: EnumInfo,
+    python_type: Optional[Type[E]] = None,
+    context: Optional[AdaptContext] = None,
+) -> None:
+    """Register the adapters to load and dump a enum type.
+
+    :param info: The object with the information about the enum to register.
+    :param python_type: Python enum type matching to the Postgres one. If `!None`,
+        the type will be generated and put into info.python_type.
+    :param context: The context where to register the adapters. If `!None`,
+        register it globally.
+
+    .. note::
+        Only string enums are supported.
+
+        Use binary format if any of your enum labels contains comma:
+            connection.execute(..., binary=True)
+    """
+
+    if not info:
+        raise TypeError("no info passed. Is the requested enum available?")
+
+    if python_type is not None:
+        if {type(item.value) for item in python_type} != {str}:
+            raise TypeError("invalid enum value type (string is the only supported)")
+
+        info.python_type = python_type
+    else:
+        info.python_type = Enum(  # type: ignore
+            info.name.title(),
+            {label: label for label in info.enum_labels},
+        )
+
+    adapters = context.adapters if context else postgres.adapters
+
+    info.register(context)
+
+    base = EnumLoader
+    name = f"{info.name.title()}{base.__name__}"
+    attribs = {"python_type": info.python_type}
+    loader = type(name, (base,), attribs)
+    adapters.register_loader(info.oid, loader)
+
+    base = EnumBinaryLoader
+    name = f"{info.name.title()}{base.__name__}"
+    attribs = {"python_type": info.python_type}
+    loader = type(name, (base,), attribs)
+    adapters.register_loader(info.oid, loader)
+
+
+def register_default_adapters(context: AdaptContext) -> None:
```

Comment:
Looks like the generic dumpers are ok. My fear was about creating an incompatibility with people dumping int enums as postgres integers. But there is a test in `test_numeric.py` proving that this use is not broken.

However, in order to work ok, the oid of the generic dumpers should be 0, or you get an error such:
```
psycopg.errors.DatatypeMismatch: column "val" is of type testenum but expression is of type text
```
added a failing test to prove it.

Reply:
What if just remove the generic dumpers? They are kinda useless, because AFAIK `StrDumperUnknown` is being used for all Python enums by default.

Reply:
Also generic dumpers breaks tests.
Some tests generates queries based on existing dumpers. There was no `make_Enum` method in Faker, but if I add it:
```python
    def make_Enum(self, spec, length=5):
        return Enum(
            f"FakeEnum{randrange(0, 1 << 90)}",
            {f"K{i}": f"V{i}" for i in range(1, randrange(1, length))},
        )
```
It still raises error:
```
psycopg.ProgrammingError: cannot adapt type 'EnumMeta' using placeholder '%t' (format: TEXT)
```
Excuse me, but I just don't know how to fix it.

Reply:
I'll look into that, but probably we will have to exclude enums from these random tests, because having them would also require to generate a type on the postgres side.

Reply:
I made a fix to disable random tests of enums.

Reply:
Thank you!